### PR TITLE
Fix silent fail in ssh_add_suseconnect_product

### PR DESCRIPTION
**Pull request by grisu48**
Fix silent failures in `ssh_add_suseconnect_product`

Currently failures in `ssh_add_suseconnect_product` are silent and remain undetected, if `$retry` is not set (by default not set).

- Related ticket: https://progress.opensuse.org/issues/69874
- Needles: -
- Verification run: http://phoenix-openqa.qam.suse.de/tests/3209

### CODE REVIEW

**Comment by dzedro** [comment]
I tried to understand what is going here. I just remembered I added the retry in `publiccloud/register_system` not knowing about retry within `ssh_add_suseconnect_product`.
Now without retry argument it fail thanks to `assert_script_run`, all good. The retry within `publiccloud/register_system` was never used, so now it will just fail, which is good, but I think we need retry otherwise we will have to restart this test every day.
With retry it will run the `$command` twice + retry and  fail at `assert_script_run` because the  retry will register the module before.
record_soft_failure will be not executed with one retry, the change is complicated with additional problems.
The problem is script_run, I would use the retry with eval which i used in `publiccloud/register_system` with `assert_script_run`.

**Comment by grisu48** [comment]
> The problem is `script_run`, I would use the retry with eval which i used in `publiccloud/register_system` with `assert_script_run`.

You mean using `script_retry` (e.g. 5 times) and grep/eval for "Successfully registered system" as it can be found in http://phoenix-openqa.qam.suse.de/tests/3172#step/register_system/25, right?

**Comment by dzedro** [comment]
> > The problem is `script_run`, I would use the retry with eval which i used in `publiccloud/register_system` with `assert_script_run`.
> 
> You mean using `script_retry` (e.g. 5 times) and grep/eval for "Successfully registered system" as it can be found in http://phoenix-openqa.qam.suse.de/tests/3172#step/register_system/25, right?

Either use script_retry or the eval.

**Review by pdostal — APPROVED** [review]
I think it's OK now :)

**Comment by dzedro** [comment]
> I think it's OK now :)

It is not! it's the same as I described before ... same problems.

**Comment by grisu48** [comment]
> It is not! it's the same as I described before ... same problems.

Help me understand the problem. In all cases we run `SUSEConnect` now 6 times with 10 seconds delay. This mitigates occasional network/scc hickups at peak times.
There is bsc#1124318 for which we try to run `SUSEConnect` once (if `$retry` is set and PackageHub is the package). If this fails we report a softfailure and then also run `script_retry`. In both cases we try to run `SUSEConnect` several times before we give up.

I thought that's what you meant, or am I missing something?

**Comment by dzedro** [comment]
> > It is not! it's the same as I described before ... same problems.
> 
> Help me understand the problem. In all cases we run `SUSEConnect` now 6 times with 10 seconds delay. This mitigates occasional network/scc hickups at peak times.
> There is bsc#1124318 for which we try to run `SUSEConnect` once (if `$retry` is set and PackageHub is the package). If this fails we report a softfailure and then also run `script_retry`. In both cases we try to run `SUSEConnect` several times before we give up.
> 
> I thought that's what you meant, or am I missing something?
> 
> IMHO If we replace `script_run` with `script_retry` then there is no need for the softfailure anymore, as it is exactly the point that the first try fails.

Sorry it's not the same, now the problems are with retry and `PackageHub`. What will happen with retry on everything else ? No retry ... ?

**Comment by dzedro** [comment]
> IMHO If we replace `script_run` with `script_retry` then there is no need for the softfailure anymore, as it is exactly the point that the first try fails.

`script_retry` does do die ... 
I agree  that all that `$retry` is not needed with `script_retry`

**Comment by grisu48** [comment]
> > IMHO If we replace `script_run` with `script_retry` then there is no need for the softfailure anymore, as it is exactly the point that the first try fails.
> 
> `script_retry` does do die ...
> I agree that all that `$retry` is not needed with `script_retry`

It looks like `$retry` is anyways not used (anymore)

    $ grep -r 'ssh_add_suseconnect_product' {tests,lib}
    ./tests/publiccloud/register_system.pm:            ssh_add_suseconnect_product($args->{my_instance}->public_ip, get_addon_fullname($addon), '`echo ${VERSION} | cut -d- -f1`') unless ($addon eq '');
    ./tests/publiccloud/register_system.pm:            ssh_add_suseconnect_product($args->{my_instance}->public_ip, get_addon_fullname($addon), '${VERSION_ID}') unless ($addon eq '');
    ./tests/publiccloud/register_system.pm:            ssh_add_suseconnect_product($args->{my_instance}->public_ip, get_addon_fullname($addon)) unless ($addon eq '');
    lib/registration.pm:  ssh_add_suseconnect_product
    lib/registration.pm:=head2 ssh_add_suseconnect_product
    lib/registration.pm:    ssh_add_suseconnect_product($remote, $name, [$version, [$arch, [$params]]]);
    lib/registration.pm:sub ssh_add_suseconnect_product {

So, I suggest we get rid of this parameter as a whole and just do a simple `script_retry`

**Comment by dzedro** [comment]
I don't think `$retry` was ever used. :roll_eyes:

**Review by dzedro — APPROVED** [review]

**Comment by grisu48** [comment]
I've added the two function `ssh_script_run` and `ssh_script_retry` to `lib/publiccloud/instance.pm`. `ssh_script_run` is just a wrapper function for `run_ssh_command` but with a much closer syntax and name to `script_run`:

    $args->{my_instance}->run_ssh_command(cmd => "uname -a")
    $args->{my_instance}->ssh_script_run("uname -a")

In addition `tests/publiccloud/ssh_interactive_init.pm` warns the user if he has cloned a job with a custom `SCHEDULE` (this is not supported and a very common pitfall)

File: lib/publiccloud/instance.pm [hunk 1]
```diff
@@ -85,6 +85,66 @@ sub run_ssh_command {
...
+Runs a C<command> via ssh in the given PC instance and returns the return code
+Timeout can be set by C<timeout> or 90 sec by default.
+By default, the command is passed in single quotes to SSH.
+To avoid quoting use C<<no_quote=>1>>.
+With C<<ssh_opts=>'...'>> you can overwrite all default ops which are in
+C<<$instance->ssh_opts>>.
+Use argument C<username> to specify a different username then
+C<<$instance->username()>>.
+
+This is a wrapper function for run_ssh_command
```
Comment (cfconrad) [inline lib/publiccloud/instance.pm:102]: If it is a wrappe, I woudn't copy the perlpod of the original function. Just pointing to it and describe the differences.

File: tests/publiccloud/ssh_interactive_init.pm [hunk 2]
```diff
@@ -19,6 +19,14 @@ use utils;
 sub run {
     my ($self, $args) = @_;
 
+    # If someone schedules a publiccloud run with a custom SCHEDULE this causes
+    # the test to break, because we need to pass $args via the main_common.pm.
+    # This check warns about this common pitfall when doing verification runs
+    if (!defined $args) {
+        record_info("Note: Running publiccloud runs with a custom SCHEDULE is not yet supported");
+        die "$args is not defined";
```
Comment (cfconrad) [inline tests/publiccloud/ssh_interactive_init.pm:27]: This will cause in a use of uninitialized string in concatenation error.

Comment (cfconrad) [inline tests/publiccloud/ssh_interactive_init.pm:27]:
```suggestion
        my $msg = 'Note: Running publiccloud runs with a custom SCHEDULE is not yet supported';
        record_info('ERROR', $msg, result=>'fail');
        die ($msg);
```

File: tests/publiccloud/register_system.pm [hunk 3]
```diff
@@ -27,16 +27,8 @@ sub run {
...
-        eval {
-            $args->{my_instance}->run_ssh_command(cmd => "sudo SUSEConnect -r " . get_required_var('SCC_REGCODE'), timeout => 420) unless (get_var('FLAVOR') =~ 'On-Demand');
-        };
-        last unless ($@);
-        diag "SUSEConnect failed: $@";
-        diag "Maybe the SCC or network is busy. Retry: $_ of $max_retries";
-    }
-    die "SCC registration on publiccloud failed (with retries)" if $@;
+    # note: ssh_script_retry dies on failure
+    $args->{my_instance}->ssh_script_retry("sudo SUSEConnect -r " . get_required_var('SCC_REGCODE'), timeout => 420, retry => 3) unless (get_var('FLAVOR') =~ 'On-Demand');
```
Comment (cfconrad) [inline tests/publiccloud/register_system.pm:31]:
```suggestion
    $args->{my_instance}->ssh_script_retry("sudo SUSEConnect -r " . get_required_var('SCC_REGCODE'), timeout => 420, retry => 3) if (get_var('FLAVOR') =~ 'BYOS');
```
I think On-Demand was renamed and we now just marking BYOS, or @pdostal @jlausuch 
also see https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/11206/files#diff-445d51faa13570063c2cc4756731d747292e18f6b3288df8b544c30f1ac2a5f7L31

File: tests/publiccloud/ssh_interactive_init.pm (hunk 2 above)
Comment (grisu48) [reply tests/publiccloud/ssh_interactive_init.pm]: Fixed

Comment (grisu48) [reply tests/publiccloud/ssh_interactive_init.pm]: Thanks, I've included your suggestion now

File: tests/publiccloud/register_system.pm (hunk 3 above)
Comment (grisu48) [reply tests/publiccloud/register_system.pm]: It's indeed better for the test coverage to do the registration only on systems, that are required to do so. I've incorporated your suggestion

File: lib/publiccloud/instance.pm (hunk 1 above)
Comment (grisu48) [reply lib/publiccloud/instance.pm]: I've removed the wrapper function and refactored the original method instead.

File: lib/publiccloud/instance.pm [hunk 4]
```diff
@@ -85,6 +85,66 @@ sub run_ssh_command {
...
+By default, the command is passed in single quotes to SSH.
+To avoid quoting use C<<no_quote=>1>>.
+With C<<ssh_opts=>'...'>> you can overwrite all default ops which are in
+C<<$instance->ssh_opts>>.
+Use argument C<username> to specify a different username then
+C<<$instance->username()>>.
+=cut
+sub ssh_script_retry {
+    my ($self, $cmd, %args) = @_;
+    $cmd = $args{cmd} if (!defined $cmd && $args{cmd});
```
[the remark this answers is not in the document]
Comment (grisu48) [reply lib/publiccloud/instance.pm:105]: I've applied your changes to `run_ssh_command` and removed the wrapper function. That was indeed unnecessary and looks IMHO better this way. What do you think about my new suggestion?

**Review by cfconrad — APPROVED** [review]
LGTM, waiting for VR's

File: lib/publiccloud/instance.pm [hunk 5]
```diff
@@ -39,15 +39,16 @@ Timeout can be set by C<timeout> or 90 sec by default.
...
 With C<<ssh_opts=>'...'>> you can overwrite all default ops which are in
 C<<$instance->ssh_opts>>.
 Use argument C<username> to specify a different username then
 C<<$instance->username()>>.
 Use argument C<rc_only> to only check for the return code of the command.
 =cut
 sub run_ssh_command {
-    my ($self, %args) = @_;
+    my $self = shift;
+    my %args = testapi::compat_args({cmd => undef}, ['cmd'], @_);
```
Comment (pdostal) [inline lib/publiccloud/instance.pm:51]: What does this mean? 😄

File: lib/publiccloud/instance.pm [hunk 6]
```diff
@@ -85,6 +86,43 @@ sub run_ssh_command {
...
+To avoid quoting use C<<no_quote=>1>>.
+With C<<ssh_opts=>'...'>> you can overwrite all default ops which are in
+C<<$instance->ssh_opts>>.
+Use argument C<username> to specify a different username then
+C<<$instance->username()>>.
+=cut
+sub ssh_script_retry {
+    my ($self, $cmd, %args) = @_;
+    $cmd = $args{cmd} if (!defined $cmd && $args{cmd});
+    $args{rc_only} = 1;
```
Comment (pdostal) [inline lib/publiccloud/instance.pm:106]: Whys is this in here? It's not used elsewhere - just defined in here.

File: lib/publiccloud/instance.pm [hunk 7]
```diff
@@ -85,6 +86,43 @@ sub run_ssh_command {
...
+        # ssh_script_run can throw a timeout exception
+        eval {
+            $rc = $self->run_ssh_command($cmd, %args);
+        };
+        if ($@ or $rc != 0) {
+            sleep($delay);
+            next;
+        } else {
+            return $rc;
+        }
```
Comment (pdostal) [inline lib/publiccloud/instance.pm:121]:
```suggestion
        my $rc = $self->run_ssh_command($cmd, %args);
        return $rc if ($rc != 0);
        sleep($delay);
```

File: lib/publiccloud/instance.pm [hunk 8]
```diff
@@ -85,6 +86,43 @@ sub run_ssh_command {
...
+    }
+    die "ssh_script_retry failed for command '$cmd'";
```
Comment (pdostal) [inline lib/publiccloud/instance.pm:123]: I'd wait for Godot instead!
```suggestion
    die "Waiting for Godot: ".$cmd;
```

File: lib/publiccloud/instance.pm [hunk 9]
```diff
@@ -85,6 +86,43 @@ sub run_ssh_command {
...
+the given number of retries is exhausted and an exception is thrown.
+Timeout can be set by C<timeout> or 90 sec by default.
+By default, the command is passed in single quotes to SSH.
+To avoid quoting use C<<no_quote=>1>>.
+With C<<ssh_opts=>'...'>> you can overwrite all default ops which are in
+C<<$instance->ssh_opts>>.
+Use argument C<username> to specify a different username then
+C<<$instance->username()>>.
+=cut
+sub ssh_script_retry {
```
Comment (pdostal) [inline lib/publiccloud/instance.pm:103]:
```suggestion
sub retry_ssh_command {
```

File: lib/registration.pm [hunk 10]
```diff
@@ -195,27 +195,22 @@ sub add_suseconnect_product {
...
+    $delay   //= 10;
 
-    my $result = script_run("ssh $remote sudo SUSEConnect -p $name/$version/$arch $params", $timeout);
-    if ($result != 0 && $retry) {
-        if ($name =~ /PackageHub/) {
-            record_soft_failure 'bsc#1124318 - Fail to get module repo metadata - running the command again as a workaround';
-        }
-        assert_script_run("ssh $remote sudo SUSEConnect -p $name/$version/$arch $params", $timeout);
-    }
+    script_retry("ssh $remote sudo SUSEConnect -p $name/$version/$arch $params", delay => $delay, retry => $retries, timeout => $timeout);
```
Comment (pdostal) [inline lib/registration.pm:213]:
```suggestion
    ssh_script_retry("ssh $remote sudo SUSEConnect -p $name/$version/$arch $params", delay => $delay, retry => $retries, timeout => $timeout);
```
or better :)
```suggestion
    retry_ssh_command("ssh $remote sudo SUSEConnect -p $name/$version/$arch $params", delay => $delay, retry => $retries, timeout => $timeout);
```

File: lib/registration.pm [hunk 11]
```diff
@@ -20,7 +20,7 @@ use Exporter;
 use strict;
 use warnings;
 use testapi;
-use utils qw(addon_decline_license assert_screen_with_soft_timeout zypper_call systemctl handle_untrusted_gpg_key common_service_start);
+use utils qw(addon_decline_license assert_screen_with_soft_timeout zypper_call systemctl handle_untrusted_gpg_key common_service_start script_retry);
```
Comment (pdostal) [inline lib/registration.pm:23]:
```suggestion
use utils qw(addon_decline_license assert_screen_with_soft_timeout zypper_call systemctl handle_untrusted_gpg_key common_service_start);
```

File: lib/publiccloud/instance.pm (hunk 5 above)
Comment (cfconrad) [reply lib/publiccloud/instance.pm:51]: this: https://github.com/os-autoinst/os-autoinst/blob/master/testapi.pm#L2243 :)

File: lib/publiccloud/instance.pm (hunk 6 above)
Comment (cfconrad) [reply lib/publiccloud/instance.pm:106]: it is used here: https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/11199/files#diff-967d56f05585b47be10c06c83e8554574ce9d259d95bac01abff8d39a7a169a2L57

but yes, myaybe the caller should set `rc_only` when calling `retry_ssh_command()`

File: lib/publiccloud/instance.pm (hunk 7 above)
Comment (cfconrad) [reply lib/publiccloud/instance.pm:121]: you would not catch SSH timeout failures... but maybe we do not want to

File: lib/publiccloud/instance.pm (hunk 8 above)
Comment (cfconrad) [reply lib/publiccloud/instance.pm:123]: https://en.wikipedia.org/wiki/Waiting_for_Godot ? :P

File: lib/publiccloud/instance.pm (hunk 6 above)
Comment (grisu48) [reply lib/publiccloud/instance.pm:106]: `rc_only` is obligatory, otherwise there is no possibility to check, if the command was successful. And this is necessary for checking, if we need to run the command again.
I would argue, that for `retry_ssh_command` this is valid because there is no way of realizing this otherwise (while sticking to the current internals).

File: lib/publiccloud/instance.pm (hunk 8 above)
Comment (pdostal) [reply lib/publiccloud/instance.pm:123]: Of course :)

File: lib/publiccloud/instance.pm (hunk 7 above)
Comment (pdostal) [reply lib/publiccloud/instance.pm:121]: There should be
```perl
return $rc if ($rc == 0);
```

Comment (grisu48) [reply lib/publiccloud/instance.pm:121]: ssh timeouts are one of the reasons why we are doing this - I could even see some in the verification runs.
This is the reason for the whole `eval` block, which indeed makes the code a bit messy here

File: lib/publiccloud/instance.pm (hunk 6 above)
Comment (pdostal) [reply lib/publiccloud/instance.pm:106]: Oh, sorry. I'd maybe make a comment about it..

File: lib/publiccloud/instance.pm (hunk 5 above)
Comment (pdostal) [reply lib/publiccloud/instance.pm:51]: And why this is better than the one-lines from before?

File: lib/publiccloud/instance.pm (hunk 9 above)
Comment (grisu48) [reply lib/publiccloud/instance.pm:103]: While your suggestion looks nicer per se, I would argue, that `ssh_script_retry` is closer to the already established `script_retry` API call and we stay somehow consistent

File: lib/publiccloud/instance.pm (hunk 8 above)
Comment (grisu48) [reply lib/publiccloud/instance.pm]: Done! :+1:

File: lib/publiccloud/instance.pm (hunk 7 above)
Comment (pdostal) [reply lib/publiccloud/instance.pm:121]: Wouldn't the suggestion I placed prevent SSH timeouts? The `return $rc if ($rc == 0);` will not fail either return but in case of SSH timeout the `for` cycle will continue.

Comment (pdostal) [reply lib/publiccloud/instance.pm:121]: And btw. if there is a timeout, we should increase it.

File: lib/publiccloud/instance.pm (hunk 9 above)
Comment (pdostal) [reply lib/publiccloud/instance.pm:103]: I also want to be consistent. And as we already have `run_ssh_command` in the same module, I'd do `retry_ssh_command`.

File: lib/publiccloud/instance.pm (hunk 7 above)
Comment (grisu48) [reply lib/publiccloud/instance.pm:121]: Timeouts are build-in into `run_ssh_command` - See https://github.com/grisu48/os-autoinst-distri-opensuse/blob/9c365be6d60a4426f24fdcb41acce20d7fefe03f/lib/publiccloud/instance.pm#L55
If the `timeout` setting is not set, it defaults to 90, which should be sufficient for most tasks (we could consider increasing this for SCC though)
Setting the `timeout` to zero will run the command without waiting and returning the return code, so that's also not an option.

In other words: If we want to keep `run_ssh_command` as it is, the only way I see is to set `rc_only`, which requires a timeout and will throw an expection on timeout. We need to account for this with the `eval` block.

File: lib/publiccloud/instance.pm (hunk 9 above)
Comment (grisu48) [reply lib/publiccloud/instance.pm:103]: I guess consistency within the module trumps over consistency with the testapi. That's ok with me :+1:

File: lib/publiccloud/instance.pm (hunk 7 above)
Comment (pdostal) [reply lib/publiccloud/instance.pm:121]: The [lib/publiccloud/instance.pm#L81](https://github.com/grisu48/os-autoinst-distri-opensuse/blob/9c365be6d60a4426f24fdcb41acce20d7fefe03f/lib/publiccloud/instance.pm#L81) should not happen in this loop. There should be something like `die => 0` option for the `run_ssh_command` routine.

Comment (grisu48) [reply lib/publiccloud/instance.pm]: Included

**Review by pdostal — APPROVED** [review]
I think now we should be ready to go 👍

**Review by pdostal — APPROVED** [review]

File: lib/publiccloud/instance.pm [hunk 12]
```diff
@@ -39,38 +39,44 @@ Timeout can be set by C<timeout> or 90 sec by default.
...
 failing (return non-zero exit code)
 By default, the command is passed in single quotes to SSH.
-To avoid quoting us C<<no_quote=>1>>.
+To avoid quoting use C<<no_quote=>1>>.
 With C<<ssh_opts=>'...'>> you can overwrite all default ops which are in
 C<<$instance->ssh_opts>>.
 Use argument C<username> to specify a different username then
 C<<$instance->username()>>.
 Use argument C<rc_only> to only check for the return code of the command.
+Use argument C<cmd_timeout> to specify a soft timeout for the command. This must be lower than C<timeot>-1 otherwise it will be set to that value.
```
Comment (cfconrad) [inline lib/publiccloud/instance.pm:48]:
```suggestion
Use argument C<cmd_timeout> to specify a soft timeout for the command. This must be lower than C<timeout>-1 otherwise it will be set to that value.
```

File: lib/publiccloud/instance.pm [hunk 13]
```diff
@@ -39,38 +39,44 @@ Timeout can be set by C<timeout> or 90 sec by default.
...
+    my %args = testapi::compat_args({cmd => undef}, ['cmd'], @_);
     die('Argument <cmd> missing') unless ($args{cmd});
     $args{ssh_opts} //= $self->ssh_opts() . " -i '" . $self->ssh_key . "'";
     $args{username} //= $self->username();
     $args{timeout}  //= SSH_TIMEOUT;
     $args{quiet}    //= 1;
     $args{no_quote} //= 0;
-    my $rc_only = $args{rc_only} // 0;
+    my $rc_only     = $args{rc_only}     // 0;
+    my $cmd_timeout = $args{cmd_timeout} // 0;
```
Comment (cfconrad) [inline lib/publiccloud/instance.pm:60]: did you found a place where we would break old code, when introduce `if ($args{timeout} > 0) { $cmd_timeout = $args{timeout}; $args{timeout}+=3;}` by default?
Or is there a other reason to not using `cmd_timeout` always?

File: lib/publiccloud/instance.pm [hunk 14]
```diff
@@ -85,6 +91,37 @@ sub run_ssh_command {
...
+C<<$instance->username()>>.
+=cut
+sub retry_ssh_command {
+    my ($self, $cmd, %args) = @_;
+    $cmd = $args{cmd} if (!defined $cmd && $args{cmd});
+    $args{rc_only} = 1;
+    $args{cmd_timeout} //= 90;    # Timeout before we cancel the command
+    $args{timeout}     //= 95;    # Hard timeout before we consider something going horribly wrong
+    my $tries = $args{retry} // 3;
+    my $delay = $args{delay} // 10;
```
Comment (cfconrad) [inline lib/publiccloud/instance.pm:115]: nit thing...
```suggestion
    my $tries = delete $args{retry} // 3;
    my $delay = delete $args{delay} // 10;
```
maybe this, as you don't wanna pass it to `run_ssh_command()` otherwise there is no need for `$delay` variable.

File: lib/registration.pm (hunk 10 above)
Comment (cfconrad) [inline lib/registration.pm:213]: Now the code looks good, I'm starting to think what it mean to do a `SUSEConnect` just twice. Should we consider to make some stats about such failures. And if it happen "always" we should raise a hand? 
For me this question is actually out of the skope of this PR, cause we need to make thinks testable. But lets discuss it in next sync meeting, wdyt?

**Review by cfconrad — APPROVED** [review]
Nit things to discuss, only the typo in docs _need_ to be changed :)
LGTM!

File: lib/publiccloud/instance.pm (hunk 13 above)
Comment (grisu48) [reply lib/publiccloud/instance.pm]: I wanted to prevent unexpected side-effects, but since you suggest to use `$args{timeout}` directly, I changes the behaviour accordingly.

File: lib/registration.pm (hunk 10 above)
Comment (grisu48) [reply lib/registration.pm:213]: The reason are occasional failures because a ton of jobs get scheduled at the same time and this puts a huge load on the scc servers. Those can be mitigated by running `SUSEConnect` multiple times with adequate timeout and delay settings and this is the whole reason for this PR.

But this is indeed perhaps food for another meeting :+1:

File: lib/publiccloud/instance.pm (hunk 4 above)
[the remark this answers is not in the document]
Comment (cfconrad) [reply lib/publiccloud/instance.pm:108]: looks very nice to me. I would just remove this line now.

**Comment by grisu48** [comment]
VR is OK. IMHO ready to merge :rocket:
